---
title: "Reading, querying and writing GWAS summary data in VCF format"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Reading, querying and writing GWAS summary data in VCF format}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

We developed a format for storing and harmonising GWAS summary data known as [GWAS VCF format](https://github.com/MRCIEU/gwas_vcf_spec). This format is effective for being very fast when querying chromosome and position ranges, handling multiallelic variants and indels. 

All the data in the [IEU GWAS database](https://gwas.mrcieu.ac.uk/) is available for download in the GWAS VCF format. This R package provides fast and convenient functions for querying and creating GWAS summary data in GWAS VCF format. The package builds on the [VariantAnnotation](https://bioconductor.org/packages/release/bioc/html/VariantAnnotation.html) Bioconductor package, which itself is based on the widely used [SummarizedExperiment](https://bioconductor.org/packages/release/bioc/html/SummarizedExperiment.html) Bioconductor package.


## External tools

For some VCF querying functions it is faster to optionally use [bcftools](https://samtools.github.io/bcftools/bcftools.html), and when available the R package will use that strategy. To set a location for the bcftools package, use

```
set_bcftools('/path/to/bcftools')
```

Note that there is bcftools binary for Windows available, so some querying options will be slower on Windows.

For LD related functions the package uses [plink 1.90](https://www.cog-genomics.org/plink/1.9). You can specify the location of your plink installation by running

```
set_plink('/path/to/plink')
```

Alternatively you can automatically use use the binaries bundled here: https://github.com/explodecomputer/genetics.binaRies 

```
devtools::install_github('explodecomputer/genetics.binaRies')
set_plink()
set_bcftools()
```

To unset a path:

```
set_plink(NULL)
set_bcftools(NULL)
```


## Reading in everything

```r
vcf <- readVcf("IEU-a-2.vcf.gz")
```


## Reading in with filters

The `query_vcf` function takes either a filename to a vcf file, or vcf object as the main argument. You can then query on `rsid`, `pval` or `chrompos`.

```r
vcf <- query_vcf("IEU-a-2.vcf.gz", rsid=c("rs3128126", "rs3121561", "rs3813193"))
vcf <- query_vcf("IEU-a-2.vcf.gz", pval=5e-8)
vcf <- query_vcf("IEU-a-2.vcf.gz", chrompos=c("1:1097291-1099437"))
```

Chaining together e.g.

```r
vcf <- query_vcf("IEU-a-2.vcf.gz", rsid=c("rs3128126", "rs3121561", "rs3813193")) %>%
    query_vcf(pval=5e-8)
```

It's possible to have multiple GWAS studies per vcf. You can specify specific GWAS studies to read in using e.g.

```r
vcf <- query_vcf("IEU-a-2.vcf.gz", rsid=c("rs3128126", "rs3121561", "rs3813193"), id="IEU-a-2")
```

## LD proxies

If a set of rsids are requested from a vcf but some are absent, a reference panel can be used to search for LD proxies, extract them, and align the effects and alleles against the original variants that were requested.

```r
vcf <- proxy_match(
    "IEU-a-2.vcf.gz", 
    rsid, 
    bfile, 
    proxies="yes", 
    tag_kb=5000, tag_nsnp=5000, tag_r2=0.6, threads=1)
```

You may also extract only the best available proxies even if the requested rsids are present, by using `proxies="only"`.

## A note about chrompos

The fastest way to query VCFs is by specifying chromosome and position. Can specify specific positions, or ranges. e.g.

```r
cp <- c("1:10000", "2:10000-20000")
```

or as a data frame

```r
cp <- data_frame(chrom=c(1,2), start=c(10000,10000), end=c(10000, 20000))
```

You can check what will be parsed out with:

```r
parse_chrompos(cp)
```

Querying by p-value or rsid is also possible but is slower as only chrompos is indexed. On Mac and Linux, rsid and p-value queries are performed by calls to bcftools. On Windows it uses VariantAnnotation directly, because bcftools binaries are not available. This is unfortunately somewhat slower. If many operations are being performed it might be faster to read in the whole dataset and perform queries that way.

## Converting from the VCF object

The VCF object is somewhat complex and you can read more about it in the 'VariantAnnotation package documentation](https://bioconductor.org/packages/release/bioc/html/VariantAnnotation.html). You can create various other formats that might be easier to use from it. For example, create a `GRanges` object which is great for fast chromosome-position operations

```r
vcf_to_granges(vcf)
```

Create a data frame (tibble, note that it will drop rsids)

```r
vcf_to_granges(vcf) %>% as_tibble
```

Create the `exposure_dat` or `outcome_dat` objects for the TwoSampleMR package

```r
format_from_vcf(vcf, "exposure")
```

or 

```r
format_from_vcf(vcf, "outcome")
```


## Creating the VCF object from a data frame

If you have GWAS summary data in a text file or data frame, this can be converted to a VCF object.

You may want to first harmonise the data so that all the non-effect alleles are aligned to the human genome reference. To do this, obtain the 1000 genomes vcf manifest (link) to use as an allele reference, and run

```r
ref <- read_reference("filename.vcf.gz")
gwas        # This needs to be in exposure_dat format
dat <- harmonise_against_ref(gwas, ref)
```

Once done, you can create the vcf

```r
dat %$% create_vcf(...)
```

