<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Reading, querying and writing GWAS summary data in VCF format • gwasvcf</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Reading, querying and writing GWAS summary data in VCF format">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">gwasvcf</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/guide.html">Reading, querying and writing GWAS summary data in VCF format</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/mrcieu/gwasvcf">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Reading, querying and writing GWAS summary data in VCF format</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/mrcieu/gwasvcf/blob/master/vignettes/guide.Rmd"><code>vignettes/guide.Rmd</code></a></small>
      <div class="hidden name"><code>guide.Rmd</code></div>

    </div>

    
    
<p>We developed a format for storing and harmonising GWAS summary data known as <a href="https://github.com/MRCIEU/gwas_vcf_spec">GWAS VCF format</a>. This format is effective for being very fast when querying chromosome and position ranges, handling multiallelic variants and indels.</p>
<p>All the data in the <a href="https://gwas.mrcieu.ac.uk/">IEU GWAS database</a> is available for download in the GWAS VCF format. This R package provides fast and convenient functions for querying and creating GWAS summary data in GWAS VCF format. The package builds on the <a href="https://bioconductor.org/packages/release/bioc/html/VariantAnnotation.html">VariantAnnotation</a> Bioconductor package, which itself is based on the widely used <a href="https://bioconductor.org/packages/release/bioc/html/SummarizedExperiment.html">SummarizedExperiment</a> Bioconductor package.</p>
<div id="external-tools" class="section level2">
<h2 class="hasAnchor">
<a href="#external-tools" class="anchor"></a>External tools</h2>
<p>For some VCF querying functions it is faster to optionally use <a href="https://samtools.github.io/bcftools/bcftools.html">bcftools</a>, and when available the R package will use that strategy. To set a location for the bcftools package, use</p>
<pre><code><a href="../reference/set_bcftools.html">set_bcftools('/path/to/bcftools')</a></code></pre>
<p>Note that there is bcftools binary for Windows available, so some querying options will be slower on Windows.</p>
<p>For LD related functions the package uses <a href="https://www.cog-genomics.org/plink/1.9">plink 1.90</a>. You can specify the location of your plink installation by running</p>
<pre><code><a href="../reference/set_plink.html">set_plink('/path/to/plink')</a></code></pre>
<p>Alternatively you can automatically use use the binaries bundled here: <a href="https://github.com/explodecomputer/genetics.binaRies" class="uri">https://github.com/explodecomputer/genetics.binaRies</a></p>
<pre><code><a href="https://rdrr.io/pkg/devtools/man/reexports.html">devtools::install_github('explodecomputer/genetics.binaRies')
set_plink()
set_bcftools()</a></code></pre>
<p>To unset a path:</p>
<pre><code><a href="../reference/set_plink.html">set_plink(NULL)
set_bcftools(NULL)</a></code></pre>
</div>
<div id="reading-in-everything" class="section level2">
<h2 class="hasAnchor">
<a href="#reading-in-everything" class="anchor"></a>Reading in everything</h2>
<pre class="sourceCode r" id="cb5"><code class="sourceCode r"><div class="sourceLine" id="cb5-1" data-line-number="1">vcf &lt;-<span class="st"> </span><span class="kw">readVcf</span>(<span class="st">"IEU-a-2.vcf.gz"</span>)</div></code></pre>
</div>
<div id="reading-in-with-filters" class="section level2">
<h2 class="hasAnchor">
<a href="#reading-in-with-filters" class="anchor"></a>Reading in with filters</h2>
<p>The <code>query_vcf</code> function takes either a filename to a vcf file, or vcf object as the main argument. You can then query on <code>rsid</code>, <code>pval</code> or <code>chrompos</code>.</p>
<pre class="sourceCode r" id="cb6"><code class="sourceCode r"><div class="sourceLine" id="cb6-1" data-line-number="1">vcf &lt;-<span class="st"> </span><span class="kw">query_vcf</span>(<span class="st">"IEU-a-2.vcf.gz"</span>, <span class="dt">rsid=</span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"rs3128126"</span>, <span class="st">"rs3121561"</span>, <span class="st">"rs3813193"</span>))</div>
<div class="sourceLine" id="cb6-2" data-line-number="2">vcf &lt;-<span class="st"> </span><span class="kw">query_vcf</span>(<span class="st">"IEU-a-2.vcf.gz"</span>, <span class="dt">pval=</span><span class="fl">5e-8</span>)</div>
<div class="sourceLine" id="cb6-3" data-line-number="3">vcf &lt;-<span class="st"> </span><span class="kw">query_vcf</span>(<span class="st">"IEU-a-2.vcf.gz"</span>, <span class="dt">chrompos=</span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"1:1097291-1099437"</span>))</div></code></pre>
<p>Chaining together e.g.</p>
<pre class="sourceCode r" id="cb7"><code class="sourceCode r"><div class="sourceLine" id="cb7-1" data-line-number="1">vcf &lt;-<span class="st"> </span><span class="kw">query_vcf</span>(<span class="st">"IEU-a-2.vcf.gz"</span>, <span class="dt">rsid=</span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"rs3128126"</span>, <span class="st">"rs3121561"</span>, <span class="st">"rs3813193"</span>)) <span class="op">%&gt;%</span>
</div>
<div class="sourceLine" id="cb7-2" data-line-number="2">
<span class="st">    </span><span class="kw">query_vcf</span>(<span class="dt">pval=</span><span class="fl">5e-8</span>)</div></code></pre>
<p>It’s possible to have multiple GWAS studies per vcf. You can specify specific GWAS studies to read in using e.g.</p>
<pre class="sourceCode r" id="cb8"><code class="sourceCode r"><div class="sourceLine" id="cb8-1" data-line-number="1">vcf &lt;-<span class="st"> </span><span class="kw">query_vcf</span>(<span class="st">"IEU-a-2.vcf.gz"</span>, <span class="dt">rsid=</span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"rs3128126"</span>, <span class="st">"rs3121561"</span>, <span class="st">"rs3813193"</span>), <span class="dt">id=</span><span class="st">"IEU-a-2"</span>)</div></code></pre>
</div>
<div id="ld-proxies" class="section level2">
<h2 class="hasAnchor">
<a href="#ld-proxies" class="anchor"></a>LD proxies</h2>
<p>If a set of rsids are requested from a vcf but some are absent, a reference panel can be used to search for LD proxies, extract them, and align the effects and alleles against the original variants that were requested.</p>
<pre class="sourceCode r" id="cb9"><code class="sourceCode r"><div class="sourceLine" id="cb9-1" data-line-number="1">vcf &lt;-<span class="st"> </span><span class="kw"><a href="../reference/proxy_match.html">proxy_match</a></span>(</div>
<div class="sourceLine" id="cb9-2" data-line-number="2">    <span class="st">"IEU-a-2.vcf.gz"</span>, </div>
<div class="sourceLine" id="cb9-3" data-line-number="3">    rsid, </div>
<div class="sourceLine" id="cb9-4" data-line-number="4">    bfile, </div>
<div class="sourceLine" id="cb9-5" data-line-number="5">    <span class="dt">proxies=</span><span class="st">"yes"</span>, </div>
<div class="sourceLine" id="cb9-6" data-line-number="6">    <span class="dt">tag_kb=</span><span class="dv">5000</span>, <span class="dt">tag_nsnp=</span><span class="dv">5000</span>, <span class="dt">tag_r2=</span><span class="fl">0.6</span>, <span class="dt">threads=</span><span class="dv">1</span>)</div></code></pre>
<p>You may also extract only the best available proxies even if the requested rsids are present, by using <code>proxies="only"</code>.</p>
</div>
<div id="a-note-about-chrompos" class="section level2">
<h2 class="hasAnchor">
<a href="#a-note-about-chrompos" class="anchor"></a>A note about chrompos</h2>
<p>The fastest way to query VCFs is by specifying chromosome and position. Can specify specific positions, or ranges. e.g.</p>
<pre class="sourceCode r" id="cb10"><code class="sourceCode r"><div class="sourceLine" id="cb10-1" data-line-number="1">cp &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"1:10000"</span>, <span class="st">"2:10000-20000"</span>)</div></code></pre>
<p>or as a data frame</p>
<pre class="sourceCode r" id="cb11"><code class="sourceCode r"><div class="sourceLine" id="cb11-1" data-line-number="1">cp &lt;-<span class="st"> </span><span class="kw">data_frame</span>(<span class="dt">chrom=</span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">1</span>,<span class="dv">2</span>), <span class="dt">start=</span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">10000</span>,<span class="dv">10000</span>), <span class="dt">end=</span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">10000</span>, <span class="dv">20000</span>))</div></code></pre>
<p>You can check what will be parsed out with:</p>
<pre class="sourceCode r" id="cb12"><code class="sourceCode r"><div class="sourceLine" id="cb12-1" data-line-number="1">
<span class="kw"><a href="../reference/parse_chrompos.html">parse_chrompos</a></span>(cp)</div></code></pre>
<p>Querying by p-value or rsid is also possible but is slower as only chrompos is indexed. On Mac and Linux, rsid and p-value queries are performed by calls to bcftools. On Windows it uses VariantAnnotation directly, because bcftools binaries are not available. This is unfortunately somewhat slower. If many operations are being performed it might be faster to read in the whole dataset and perform queries that way.</p>
</div>
<div id="converting-from-the-vcf-object" class="section level2">
<h2 class="hasAnchor">
<a href="#converting-from-the-vcf-object" class="anchor"></a>Converting from the VCF object</h2>
<p>The VCF object is somewhat complex and you can read more about it in the ’VariantAnnotation package documentation](<a href="https://bioconductor.org/packages/release/bioc/html/VariantAnnotation.html" class="uri">https://bioconductor.org/packages/release/bioc/html/VariantAnnotation.html</a>). You can create various other formats that might be easier to use from it. For example, create a <code>GRanges</code> object which is great for fast chromosome-position operations</p>
<pre class="sourceCode r" id="cb13"><code class="sourceCode r"><div class="sourceLine" id="cb13-1" data-line-number="1">
<span class="kw"><a href="../reference/vcf_to_granges.html">vcf_to_granges</a></span>(vcf)</div></code></pre>
<p>Create a data frame (tibble, note that it will drop rsids)</p>
<pre class="sourceCode r" id="cb14"><code class="sourceCode r"><div class="sourceLine" id="cb14-1" data-line-number="1">
<span class="kw"><a href="../reference/vcf_to_granges.html">vcf_to_granges</a></span>(vcf) <span class="op">%&gt;%</span><span class="st"> </span>as_tibble</div></code></pre>
<p>Create the <code>exposure_dat</code> or <code>outcome_dat</code> objects for the TwoSampleMR package</p>
<pre class="sourceCode r" id="cb15"><code class="sourceCode r"><div class="sourceLine" id="cb15-1" data-line-number="1">
<span class="kw">format_from_vcf</span>(vcf, <span class="st">"exposure"</span>)</div></code></pre>
<p>or</p>
<pre class="sourceCode r" id="cb16"><code class="sourceCode r"><div class="sourceLine" id="cb16-1" data-line-number="1">
<span class="kw">format_from_vcf</span>(vcf, <span class="st">"outcome"</span>)</div></code></pre>
</div>
<div id="creating-the-vcf-object-from-a-data-frame" class="section level2">
<h2 class="hasAnchor">
<a href="#creating-the-vcf-object-from-a-data-frame" class="anchor"></a>Creating the VCF object from a data frame</h2>
<p>If you have GWAS summary data in a text file or data frame, this can be converted to a VCF object.</p>
<p>You may want to first harmonise the data so that all the non-effect alleles are aligned to the human genome reference. To do this, obtain the 1000 genomes vcf manifest (link) to use as an allele reference, and run</p>
<pre class="sourceCode r" id="cb17"><code class="sourceCode r"><div class="sourceLine" id="cb17-1" data-line-number="1">ref &lt;-<span class="st"> </span><span class="kw"><a href="../reference/read_reference.html">read_reference</a></span>(<span class="st">"filename.vcf.gz"</span>)</div>
<div class="sourceLine" id="cb17-2" data-line-number="2">gwas        <span class="co"># This needs to be in exposure_dat format</span>
</div>
<div class="sourceLine" id="cb17-3" data-line-number="3">dat &lt;-<span class="st"> </span><span class="kw"><a href="../reference/harmonise_against_ref.html">harmonise_against_ref</a></span>(gwas, ref)</div></code></pre>
<p>Once done, you can create the vcf</p>
<pre class="sourceCode r" id="cb18"><code class="sourceCode r"><div class="sourceLine" id="cb18-1" data-line-number="1">dat <span class="op">%$%</span><span class="st"> </span><span class="kw"><a href="../reference/create_vcf.html">create_vcf</a></span>(...)</div></code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#external-tools">External tools</a></li>
      <li><a href="#reading-in-everything">Reading in everything</a></li>
      <li><a href="#reading-in-with-filters">Reading in with filters</a></li>
      <li><a href="#ld-proxies">LD proxies</a></li>
      <li><a href="#a-note-about-chrompos">A note about chrompos</a></li>
      <li><a href="#converting-from-the-vcf-object">Converting from the VCF object</a></li>
      <li><a href="#creating-the-vcf-object-from-a-data-frame">Creating the VCF object from a data frame</a></li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Gibran Hemani.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
