<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Reading, querying and writing GWAS summary data in VCF format • gwasvcf</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Reading, querying and writing GWAS summary data in VCF format">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">gwasvcf</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/guide.html">Reading, querying and writing GWAS summary data in VCF format</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/mrcieu/gwasvcf">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Reading, querying and writing GWAS summary data in VCF format</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/mrcieu/gwasvcf/blob/master/vignettes/guide.Rmd"><code>vignettes/guide.Rmd</code></a></small>
      <div class="hidden name"><code>guide.Rmd</code></div>

    </div>

    
    
<p>We developed a format for storing and harmonising GWAS summary data known as <a href="https://github.com/MRCIEU/gwas_vcf_spec">GWAS VCF format</a>. This format is effective for being very fast when querying chromosome and position ranges, handling multiallelic variants and indels.</p>
<p>All the data in the <a href="https://gwas.mrcieu.ac.uk/">IEU GWAS database</a> is available for download in the GWAS VCF format. This R package provides fast and convenient functions for querying and creating GWAS summary data in GWAS VCF format. The package builds on the <a href="https://bioconductor.org/packages/release/bioc/html/VariantAnnotation.html">VariantAnnotation</a> Bioconductor package, which itself is based on the widely used <a href="https://bioconductor.org/packages/release/bioc/html/SummarizedExperiment.html">SummarizedExperiment</a> Bioconductor package.</p>
<div id="external-tools" class="section level2">
<h2 class="hasAnchor">
<a href="#external-tools" class="anchor"></a>External tools</h2>
<p>For some VCF querying functions it is faster to optionally use <a href="https://samtools.github.io/bcftools/bcftools.html">bcftools</a>, and when available the R package will use that strategy. To set a location for the bcftools package, use</p>
<pre><code><a href="../reference/set_bcftools.html">set_bcftools('/path/to/bcftools')</a></code></pre>
<p>Note that there is bcftools binary for Windows available, so some querying options will be slower on Windows.</p>
<p>For LD related functions the package uses <a href="https://www.cog-genomics.org/plink/1.9">plink 1.90</a>. You can specify the location of your plink installation by running</p>
<pre><code><a href="../reference/set_plink.html">set_plink('/path/to/plink')</a></code></pre>
<p>Alternatively you can automatically use use the binaries bundled here: <a href="https://github.com/explodecomputer/genetics.binaRies" class="uri">https://github.com/explodecomputer/genetics.binaRies</a></p>
<pre><code><a href="https://rdrr.io/pkg/devtools/man/remote-reexports.html">devtools::install_github('explodecomputer/genetics.binaRies')
set_plink()
set_bcftools()</a></code></pre>
<p>To unset a path:</p>
<pre><code><a href="../reference/set_plink.html">set_plink(NULL)
set_bcftools(NULL)</a></code></pre>
<p>For this vignette we will use the bundled binaries in <code>genetics.binaRies</code>.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a><span class="kw"><a href="https://rdrr.io/r/base/warning.html">suppressWarnings</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/message.html">suppressPackageStartupMessages</a></span>({</span>
<span id="cb5-2"><a href="#cb5-2"></a>  <span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(gwasvcf)</span>
<span id="cb5-3"><a href="#cb5-3"></a>  <span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(VariantAnnotation)</span>
<span id="cb5-4"><a href="#cb5-4"></a>  <span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(dplyr)</span>
<span id="cb5-5"><a href="#cb5-5"></a>  <span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(magrittr)</span>
<span id="cb5-6"><a href="#cb5-6"></a>}))</span>
<span id="cb5-7"><a href="#cb5-7"></a><span class="kw"><a href="../reference/set_bcftools.html">set_bcftools</a></span>()</span>
<span id="cb5-8"><a href="#cb5-8"></a><span class="co">#&gt; Loading required namespace: genetics.binaRies</span></span>
<span id="cb5-9"><a href="#cb5-9"></a><span class="co">#&gt; Path not provided, using binaries in the explodecomputer/genetics.binaRies package</span></span></code></pre></div>
</div>
<div id="reading-in-everything" class="section level2">
<h2 class="hasAnchor">
<a href="#reading-in-everything" class="anchor"></a>Reading in everything</h2>
<p>To read an entire dataset use the <code>readVcf</code> function. As an example we’ll use the bundled data which is a small subset of the Speliotes et al 2010 BMI GWAS.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1"></a>vcffile &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"data.vcf.gz"</span>, <span class="dt">package=</span><span class="st">"gwasvcf"</span>)</span>
<span id="cb6-2"><a href="#cb6-2"></a>vcf &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/VariantAnnotation/man/readVcf-methods.html">readVcf</a></span>(vcffile)</span>
<span id="cb6-3"><a href="#cb6-3"></a><span class="kw"><a href="https://rdrr.io/r/base/class.html">class</a></span>(vcf)</span>
<span id="cb6-4"><a href="#cb6-4"></a><span class="co">#&gt; [1] "CollapsedVCF"</span></span>
<span id="cb6-5"><a href="#cb6-5"></a><span class="co">#&gt; attr(,"package")</span></span>
<span id="cb6-6"><a href="#cb6-6"></a><span class="co">#&gt; [1] "VariantAnnotation"</span></span></code></pre></div>
<p>Please refer to the <code>VariantAnnotation</code> package documentation for full details about the <code>CollapsedVCF</code> object. A brief summary follows.</p>
<p>General info about the dataset can be obtained by calling it:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a>vcf</span>
<span id="cb7-2"><a href="#cb7-2"></a><span class="co">#&gt; class: CollapsedVCF </span></span>
<span id="cb7-3"><a href="#cb7-3"></a><span class="co">#&gt; dim: 92 1 </span></span>
<span id="cb7-4"><a href="#cb7-4"></a><span class="co">#&gt; rowRanges(vcf):</span></span>
<span id="cb7-5"><a href="#cb7-5"></a><span class="co">#&gt;   GRanges with 5 metadata columns: paramRangeID, REF, ALT, QUAL, FILTER</span></span>
<span id="cb7-6"><a href="#cb7-6"></a><span class="co">#&gt; info(vcf):</span></span>
<span id="cb7-7"><a href="#cb7-7"></a><span class="co">#&gt;   DataFrame with 1 column: AF</span></span>
<span id="cb7-8"><a href="#cb7-8"></a><span class="co">#&gt; info(header(vcf)):</span></span>
<span id="cb7-9"><a href="#cb7-9"></a><span class="co">#&gt;       Number Type  Description     </span></span>
<span id="cb7-10"><a href="#cb7-10"></a><span class="co">#&gt;    AF A      Float Allele Frequency</span></span>
<span id="cb7-11"><a href="#cb7-11"></a><span class="co">#&gt; geno(vcf):</span></span>
<span id="cb7-12"><a href="#cb7-12"></a><span class="co">#&gt;   SimpleList of length 9: ES, SE, LP, AF, SS, EZ, SI, NC, ID</span></span>
<span id="cb7-13"><a href="#cb7-13"></a><span class="co">#&gt; geno(header(vcf)):</span></span>
<span id="cb7-14"><a href="#cb7-14"></a><span class="co">#&gt;       Number Type   Description                                           </span></span>
<span id="cb7-15"><a href="#cb7-15"></a><span class="co">#&gt;    ES A      Float  Effect size estimate relative to the alternative al...</span></span>
<span id="cb7-16"><a href="#cb7-16"></a><span class="co">#&gt;    SE A      Float  Standard error of effect size estimate                </span></span>
<span id="cb7-17"><a href="#cb7-17"></a><span class="co">#&gt;    LP A      Float  -log10 p-value for effect estimate                    </span></span>
<span id="cb7-18"><a href="#cb7-18"></a><span class="co">#&gt;    AF A      Float  Alternate allele frequency in the association study   </span></span>
<span id="cb7-19"><a href="#cb7-19"></a><span class="co">#&gt;    SS A      Float  Sample size used to estimate genetic effect           </span></span>
<span id="cb7-20"><a href="#cb7-20"></a><span class="co">#&gt;    EZ A      Float  Z-score provided if it was used to derive the EFFEC...</span></span>
<span id="cb7-21"><a href="#cb7-21"></a><span class="co">#&gt;    SI A      Float  Accuracy score of summary data imputation             </span></span>
<span id="cb7-22"><a href="#cb7-22"></a><span class="co">#&gt;    NC A      Float  Number of cases used to estimate genetic effect       </span></span>
<span id="cb7-23"><a href="#cb7-23"></a><span class="co">#&gt;    ID 1      String Study variant identifier</span></span></code></pre></div>
<p>There are 92 rows and 1 column which means 92 SNPs and one GWAS. See the header information:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1"></a><span class="kw"><a href="https://rdrr.io/pkg/VariantAnnotation/man/VCFHeader-class.html">header</a></span>(vcf)</span>
<span id="cb8-2"><a href="#cb8-2"></a><span class="co">#&gt; class: VCFHeader </span></span>
<span id="cb8-3"><a href="#cb8-3"></a><span class="co">#&gt; samples(1): IEU-a-2</span></span>
<span id="cb8-4"><a href="#cb8-4"></a><span class="co">#&gt; meta(4): fileformat META SAMPLE contig</span></span>
<span id="cb8-5"><a href="#cb8-5"></a><span class="co">#&gt; fixed(1): FILTER</span></span>
<span id="cb8-6"><a href="#cb8-6"></a><span class="co">#&gt; info(1): AF</span></span>
<span id="cb8-7"><a href="#cb8-7"></a><span class="co">#&gt; geno(9): ES SE ... NC ID</span></span></code></pre></div>
<p>See the names of the GWAS datasets (in this case just one, and it refers to the IEU GWAS database ID name):</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1"></a><span class="kw"><a href="https://rdrr.io/pkg/VariantAnnotation/man/VCFHeader-class.html">samples</a></span>(<span class="kw"><a href="https://rdrr.io/pkg/VariantAnnotation/man/VCFHeader-class.html">header</a></span>(vcf))</span>
<span id="cb9-2"><a href="#cb9-2"></a><span class="co">#&gt; [1] "IEU-a-2"</span></span></code></pre></div>
<p>In this case you can obtain information about this study through the <code>ieugwasr</code> package e.g. <code><a href="https://rdrr.io/pkg/ieugwasr/man/gwasinfo.html">ieugwasr::gwasinfo("IEU-a-2")</a></code>.</p>
<p>There are a few components within the object:</p>
<ul>
<li>
<code>header</code> which has the meta data describing the dataset, including the association result variables</li>
<li>
<code>rowRanges</code> which is information about each variant</li>
<li>
<code>info</code> which is further metadata about each variant</li>
<li>
<code>geno</code> which is the actual association results for each GWAS</li>
</ul>
<p>the <code>rowRanges</code> object is a <code>GenomicRanges</code> class, which is useful for performing fast operations on chromosome position information.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1"></a><span class="kw">rowRanges</span>(vcf)</span>
<span id="cb10-2"><a href="#cb10-2"></a><span class="co">#&gt; GRanges object with 92 ranges and 5 metadata columns:</span></span>
<span id="cb10-3"><a href="#cb10-3"></a><span class="co">#&gt;              seqnames    ranges strand | paramRangeID            REF</span></span>
<span id="cb10-4"><a href="#cb10-4"></a><span class="co">#&gt;                 &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;factor&gt; &lt;DNAStringSet&gt;</span></span>
<span id="cb10-5"><a href="#cb10-5"></a><span class="co">#&gt;   rs12565286        1    721290      * |         &lt;NA&gt;              G</span></span>
<span id="cb10-6"><a href="#cb10-6"></a><span class="co">#&gt;   rs11804171        1    723819      * |         &lt;NA&gt;              T</span></span>
<span id="cb10-7"><a href="#cb10-7"></a><span class="co">#&gt;    rs2977670        1    723891      * |         &lt;NA&gt;              G</span></span>
<span id="cb10-8"><a href="#cb10-8"></a><span class="co">#&gt;    rs3094315        1    752566      * |         &lt;NA&gt;              G</span></span>
<span id="cb10-9"><a href="#cb10-9"></a><span class="co">#&gt;    rs2073813        1    753541      * |         &lt;NA&gt;              G</span></span>
<span id="cb10-10"><a href="#cb10-10"></a><span class="co">#&gt;          ...      ...       ...    ... .          ...            ...</span></span>
<span id="cb10-11"><a href="#cb10-11"></a><span class="co">#&gt;     rs715643        1   1172907      * |         &lt;NA&gt;              C</span></span>
<span id="cb10-12"><a href="#cb10-12"></a><span class="co">#&gt;    rs6675798        1   1176597      * |         &lt;NA&gt;              T</span></span>
<span id="cb10-13"><a href="#cb10-13"></a><span class="co">#&gt;    rs6603783        1   1181751      * |         &lt;NA&gt;              T</span></span>
<span id="cb10-14"><a href="#cb10-14"></a><span class="co">#&gt;    rs6603785        1   1186502      * |         &lt;NA&gt;              A</span></span>
<span id="cb10-15"><a href="#cb10-15"></a><span class="co">#&gt;    rs6603787        1   1188225      * |         &lt;NA&gt;              G</span></span>
<span id="cb10-16"><a href="#cb10-16"></a><span class="co">#&gt;                             ALT      QUAL      FILTER</span></span>
<span id="cb10-17"><a href="#cb10-17"></a><span class="co">#&gt;              &lt;DNAStringSetList&gt; &lt;numeric&gt; &lt;character&gt;</span></span>
<span id="cb10-18"><a href="#cb10-18"></a><span class="co">#&gt;   rs12565286                  C      &lt;NA&gt;        PASS</span></span>
<span id="cb10-19"><a href="#cb10-19"></a><span class="co">#&gt;   rs11804171                  A      &lt;NA&gt;        PASS</span></span>
<span id="cb10-20"><a href="#cb10-20"></a><span class="co">#&gt;    rs2977670                  C      &lt;NA&gt;        PASS</span></span>
<span id="cb10-21"><a href="#cb10-21"></a><span class="co">#&gt;    rs3094315                  A      &lt;NA&gt;        PASS</span></span>
<span id="cb10-22"><a href="#cb10-22"></a><span class="co">#&gt;    rs2073813                  A      &lt;NA&gt;        PASS</span></span>
<span id="cb10-23"><a href="#cb10-23"></a><span class="co">#&gt;          ...                ...       ...         ...</span></span>
<span id="cb10-24"><a href="#cb10-24"></a><span class="co">#&gt;     rs715643                  T      &lt;NA&gt;        PASS</span></span>
<span id="cb10-25"><a href="#cb10-25"></a><span class="co">#&gt;    rs6675798                  C      &lt;NA&gt;        PASS</span></span>
<span id="cb10-26"><a href="#cb10-26"></a><span class="co">#&gt;    rs6603783                  C      &lt;NA&gt;        PASS</span></span>
<span id="cb10-27"><a href="#cb10-27"></a><span class="co">#&gt;    rs6603785                  T      &lt;NA&gt;        PASS</span></span>
<span id="cb10-28"><a href="#cb10-28"></a><span class="co">#&gt;    rs6603787                  T      &lt;NA&gt;        PASS</span></span>
<span id="cb10-29"><a href="#cb10-29"></a><span class="co">#&gt;   -------</span></span>
<span id="cb10-30"><a href="#cb10-30"></a><span class="co">#&gt;   seqinfo: 84 sequences from GRCh37 genome</span></span></code></pre></div>
</div>
<div id="converting-to-simple-dataframes" class="section level2">
<h2 class="hasAnchor">
<a href="#converting-to-simple-dataframes" class="anchor"></a>Converting to simple dataframes</h2>
<p>The VCF object is somewhat complex and you can read more about it in the ’VariantAnnotation package documentation](<a href="https://bioconductor.org/packages/release/bioc/html/VariantAnnotation.html" class="uri">https://bioconductor.org/packages/release/bioc/html/VariantAnnotation.html</a>). You can create various other formats that might be easier to use from it. For example, create a <code>GRanges</code> object which is great for fast chromosome-position operations</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1"></a><span class="kw"><a href="../reference/vcf_to_granges.html">vcf_to_granges</a></span>(vcf)</span>
<span id="cb11-2"><a href="#cb11-2"></a><span class="co">#&gt; GRanges object with 92 ranges and 15 metadata columns:</span></span>
<span id="cb11-3"><a href="#cb11-3"></a><span class="co">#&gt;              seqnames    ranges strand | paramRangeID            REF</span></span>
<span id="cb11-4"><a href="#cb11-4"></a><span class="co">#&gt;                 &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;factor&gt; &lt;DNAStringSet&gt;</span></span>
<span id="cb11-5"><a href="#cb11-5"></a><span class="co">#&gt;   rs12565286        1    721290      * |         &lt;NA&gt;              G</span></span>
<span id="cb11-6"><a href="#cb11-6"></a><span class="co">#&gt;   rs11804171        1    723819      * |         &lt;NA&gt;              T</span></span>
<span id="cb11-7"><a href="#cb11-7"></a><span class="co">#&gt;    rs2977670        1    723891      * |         &lt;NA&gt;              G</span></span>
<span id="cb11-8"><a href="#cb11-8"></a><span class="co">#&gt;    rs3094315        1    752566      * |         &lt;NA&gt;              G</span></span>
<span id="cb11-9"><a href="#cb11-9"></a><span class="co">#&gt;    rs2073813        1    753541      * |         &lt;NA&gt;              G</span></span>
<span id="cb11-10"><a href="#cb11-10"></a><span class="co">#&gt;          ...      ...       ...    ... .          ...            ...</span></span>
<span id="cb11-11"><a href="#cb11-11"></a><span class="co">#&gt;     rs715643        1   1172907      * |         &lt;NA&gt;              C</span></span>
<span id="cb11-12"><a href="#cb11-12"></a><span class="co">#&gt;    rs6675798        1   1176597      * |         &lt;NA&gt;              T</span></span>
<span id="cb11-13"><a href="#cb11-13"></a><span class="co">#&gt;    rs6603783        1   1181751      * |         &lt;NA&gt;              T</span></span>
<span id="cb11-14"><a href="#cb11-14"></a><span class="co">#&gt;    rs6603785        1   1186502      * |         &lt;NA&gt;              A</span></span>
<span id="cb11-15"><a href="#cb11-15"></a><span class="co">#&gt;    rs6603787        1   1188225      * |         &lt;NA&gt;              G</span></span>
<span id="cb11-16"><a href="#cb11-16"></a><span class="co">#&gt;                         ALT      QUAL      FILTER        ES        SE</span></span>
<span id="cb11-17"><a href="#cb11-17"></a><span class="co">#&gt;              &lt;DNAStringSet&gt; &lt;numeric&gt; &lt;character&gt; &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span id="cb11-18"><a href="#cb11-18"></a><span class="co">#&gt;   rs12565286              C      &lt;NA&gt;        PASS   -0.0067    0.0145</span></span>
<span id="cb11-19"><a href="#cb11-19"></a><span class="co">#&gt;   rs11804171              A      &lt;NA&gt;        PASS   -0.0146    0.0175</span></span>
<span id="cb11-20"><a href="#cb11-20"></a><span class="co">#&gt;    rs2977670              C      &lt;NA&gt;        PASS    0.0044    0.0184</span></span>
<span id="cb11-21"><a href="#cb11-21"></a><span class="co">#&gt;    rs3094315              A      &lt;NA&gt;        PASS     0.006    0.0065</span></span>
<span id="cb11-22"><a href="#cb11-22"></a><span class="co">#&gt;    rs2073813              A      &lt;NA&gt;        PASS    0.0035    0.0102</span></span>
<span id="cb11-23"><a href="#cb11-23"></a><span class="co">#&gt;          ...            ...       ...         ...       ...       ...</span></span>
<span id="cb11-24"><a href="#cb11-24"></a><span class="co">#&gt;     rs715643              T      &lt;NA&gt;        PASS    0.0019    0.0118</span></span>
<span id="cb11-25"><a href="#cb11-25"></a><span class="co">#&gt;    rs6675798              C      &lt;NA&gt;        PASS   -0.0013    0.0067</span></span>
<span id="cb11-26"><a href="#cb11-26"></a><span class="co">#&gt;    rs6603783              C      &lt;NA&gt;        PASS    -2e-04    0.0069</span></span>
<span id="cb11-27"><a href="#cb11-27"></a><span class="co">#&gt;    rs6603785              T      &lt;NA&gt;        PASS    0.0075    0.0104</span></span>
<span id="cb11-28"><a href="#cb11-28"></a><span class="co">#&gt;    rs6603787              T      &lt;NA&gt;        PASS    0.0025    0.0089</span></span>
<span id="cb11-29"><a href="#cb11-29"></a><span class="co">#&gt;                     LP        AF        SS        EZ        SI        NC</span></span>
<span id="cb11-30"><a href="#cb11-30"></a><span class="co">#&gt;              &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span id="cb11-31"><a href="#cb11-31"></a><span class="co">#&gt;   rs12565286  0.193006    0.9322    109823      &lt;NA&gt;      &lt;NA&gt;      &lt;NA&gt;</span></span>
<span id="cb11-32"><a href="#cb11-32"></a><span class="co">#&gt;   rs11804171  0.393511   0.96296     84828      &lt;NA&gt;      &lt;NA&gt;      &lt;NA&gt;</span></span>
<span id="cb11-33"><a href="#cb11-33"></a><span class="co">#&gt;    rs2977670 0.0909791   0.07143   68458.9      &lt;NA&gt;      &lt;NA&gt;      &lt;NA&gt;</span></span>
<span id="cb11-34"><a href="#cb11-34"></a><span class="co">#&gt;    rs3094315   0.44855    0.1552    131544      &lt;NA&gt;      &lt;NA&gt;      &lt;NA&gt;</span></span>
<span id="cb11-35"><a href="#cb11-35"></a><span class="co">#&gt;    rs2073813  0.135786      &lt;NA&gt;   64351.3      &lt;NA&gt;      &lt;NA&gt;      &lt;NA&gt;</span></span>
<span id="cb11-36"><a href="#cb11-36"></a><span class="co">#&gt;          ...       ...       ...       ...       ...       ...       ...</span></span>
<span id="cb11-37"><a href="#cb11-37"></a><span class="co">#&gt;     rs715643 0.0594337   0.90833    121822      &lt;NA&gt;      &lt;NA&gt;      &lt;NA&gt;</span></span>
<span id="cb11-38"><a href="#cb11-38"></a><span class="co">#&gt;    rs6675798  0.072527    0.8917    223475      &lt;NA&gt;      &lt;NA&gt;      &lt;NA&gt;</span></span>
<span id="cb11-39"><a href="#cb11-39"></a><span class="co">#&gt;    rs6603783 0.0101499       0.9    220022      &lt;NA&gt;      &lt;NA&gt;      &lt;NA&gt;</span></span>
<span id="cb11-40"><a href="#cb11-40"></a><span class="co">#&gt;    rs6603785  0.327164   0.91667    165964      &lt;NA&gt;      &lt;NA&gt;      &lt;NA&gt;</span></span>
<span id="cb11-41"><a href="#cb11-41"></a><span class="co">#&gt;    rs6603787  0.108574      &lt;NA&gt;    199099      &lt;NA&gt;      &lt;NA&gt;      &lt;NA&gt;</span></span>
<span id="cb11-42"><a href="#cb11-42"></a><span class="co">#&gt;                       ID          id</span></span>
<span id="cb11-43"><a href="#cb11-43"></a><span class="co">#&gt;              &lt;character&gt; &lt;character&gt;</span></span>
<span id="cb11-44"><a href="#cb11-44"></a><span class="co">#&gt;   rs12565286  rs12565286     IEU-a-2</span></span>
<span id="cb11-45"><a href="#cb11-45"></a><span class="co">#&gt;   rs11804171  rs11804171     IEU-a-2</span></span>
<span id="cb11-46"><a href="#cb11-46"></a><span class="co">#&gt;    rs2977670   rs2977670     IEU-a-2</span></span>
<span id="cb11-47"><a href="#cb11-47"></a><span class="co">#&gt;    rs3094315   rs3094315     IEU-a-2</span></span>
<span id="cb11-48"><a href="#cb11-48"></a><span class="co">#&gt;    rs2073813   rs2073813     IEU-a-2</span></span>
<span id="cb11-49"><a href="#cb11-49"></a><span class="co">#&gt;          ...         ...         ...</span></span>
<span id="cb11-50"><a href="#cb11-50"></a><span class="co">#&gt;     rs715643    rs715643     IEU-a-2</span></span>
<span id="cb11-51"><a href="#cb11-51"></a><span class="co">#&gt;    rs6675798   rs6675798     IEU-a-2</span></span>
<span id="cb11-52"><a href="#cb11-52"></a><span class="co">#&gt;    rs6603783   rs6603783     IEU-a-2</span></span>
<span id="cb11-53"><a href="#cb11-53"></a><span class="co">#&gt;    rs6603785   rs6603785     IEU-a-2</span></span>
<span id="cb11-54"><a href="#cb11-54"></a><span class="co">#&gt;    rs6603787   rs6603787     IEU-a-2</span></span>
<span id="cb11-55"><a href="#cb11-55"></a><span class="co">#&gt;   -------</span></span>
<span id="cb11-56"><a href="#cb11-56"></a><span class="co">#&gt;   seqinfo: 84 sequences from GRCh37 genome</span></span></code></pre></div>
<p>Create a data frame:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1"></a><span class="kw"><a href="../reference/vcf_to_granges.html">vcf_to_granges</a></span>(vcf) <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/reexports.html">as_tibble</a></span>()</span>
<span id="cb12-2"><a href="#cb12-2"></a><span class="co">#&gt; # A tibble: 92 x 20</span></span>
<span id="cb12-3"><a href="#cb12-3"></a><span class="co">#&gt;    seqnames  start    end width strand paramRangeID REF   ALT    QUAL</span></span>
<span id="cb12-4"><a href="#cb12-4"></a><span class="co">#&gt;    &lt;fct&gt;     &lt;int&gt;  &lt;int&gt; &lt;int&gt; &lt;fct&gt;  &lt;fct&gt;        &lt;chr&gt; &lt;chr&gt; &lt;dbl&gt;</span></span>
<span id="cb12-5"><a href="#cb12-5"></a><span class="co">#&gt;  1 1        721290 721290     1 *      &lt;NA&gt;         G     C        NA</span></span>
<span id="cb12-6"><a href="#cb12-6"></a><span class="co">#&gt;  2 1        723819 723819     1 *      &lt;NA&gt;         T     A        NA</span></span>
<span id="cb12-7"><a href="#cb12-7"></a><span class="co">#&gt;  3 1        723891 723891     1 *      &lt;NA&gt;         G     C        NA</span></span>
<span id="cb12-8"><a href="#cb12-8"></a><span class="co">#&gt;  4 1        752566 752566     1 *      &lt;NA&gt;         G     A        NA</span></span>
<span id="cb12-9"><a href="#cb12-9"></a><span class="co">#&gt;  5 1        753541 753541     1 *      &lt;NA&gt;         G     A        NA</span></span>
<span id="cb12-10"><a href="#cb12-10"></a><span class="co">#&gt;  6 1        754192 754192     1 *      &lt;NA&gt;         A     G        NA</span></span>
<span id="cb12-11"><a href="#cb12-11"></a><span class="co">#&gt;  7 1        768448 768448     1 *      &lt;NA&gt;         G     A        NA</span></span>
<span id="cb12-12"><a href="#cb12-12"></a><span class="co">#&gt;  8 1        775659 775659     1 *      &lt;NA&gt;         A     G        NA</span></span>
<span id="cb12-13"><a href="#cb12-13"></a><span class="co">#&gt;  9 1        777122 777122     1 *      &lt;NA&gt;         A     T        NA</span></span>
<span id="cb12-14"><a href="#cb12-14"></a><span class="co">#&gt; 10 1        779322 779322     1 *      &lt;NA&gt;         A     G        NA</span></span>
<span id="cb12-15"><a href="#cb12-15"></a><span class="co">#&gt; # … with 82 more rows, and 11 more variables: FILTER &lt;chr&gt;, ES &lt;dbl&gt;,</span></span>
<span id="cb12-16"><a href="#cb12-16"></a><span class="co">#&gt; #   SE &lt;dbl&gt;, LP &lt;dbl&gt;, AF &lt;dbl&gt;, SS &lt;dbl&gt;, EZ &lt;dbl&gt;, SI &lt;dbl&gt;, NC &lt;dbl&gt;,</span></span>
<span id="cb12-17"><a href="#cb12-17"></a><span class="co">#&gt; #   ID &lt;chr&gt;, id &lt;chr&gt;</span></span></code></pre></div>
<p>The direct conversion to formats for tools such as TwoSampleMR, coloc, and many others can also be made using the <a href="https://github.com/mrcieu/gwasglue">https://github.com/mrcieu/gwasglue</a> R package.</p>
</div>
<div id="reading-in-with-filters" class="section level2">
<h2 class="hasAnchor">
<a href="#reading-in-with-filters" class="anchor"></a>Reading in with filters</h2>
<p>The <code>query_gwas</code> function takes either a filename to a vcf file, or vcf object as the main argument. You can then query on <code>rsid</code>, <code>pval</code> or <code>chrompos</code>. For example</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1"></a>vcfsubset &lt;-<span class="st"> </span><span class="kw"><a href="../reference/query_gwas.html">query_gwas</a></span>(vcffile, <span class="dt">chrompos=</span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"1:1097291-1099437"</span>))</span></code></pre></div>
<p>and</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1"></a>vcf &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/VariantAnnotation/man/readVcf-methods.html">readVcf</a></span>(vcffile)</span>
<span id="cb14-2"><a href="#cb14-2"></a>vcfsubset &lt;-<span class="st"> </span><span class="kw"><a href="../reference/query_gwas.html">query_gwas</a></span>(vcf, <span class="dt">chrompos=</span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"1:1097291-1099437"</span>))</span></code></pre></div>
<p>are each identical, but the former saves time and memory because it is querying the file using an index and only reading in what is required.</p>
<p>Examples of other filters are here:</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1"></a>vcf &lt;-<span class="st"> </span><span class="kw"><a href="../reference/query_gwas.html">query_gwas</a></span>(vcffile, <span class="dt">rsid=</span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"rs3128126"</span>, <span class="st">"rs3121561"</span>, <span class="st">"rs3813193"</span>))</span>
<span id="cb15-2"><a href="#cb15-2"></a>vcf</span>
<span id="cb15-3"><a href="#cb15-3"></a><span class="co">#&gt; class: CollapsedVCF </span></span>
<span id="cb15-4"><a href="#cb15-4"></a><span class="co">#&gt; dim: 3 1 </span></span>
<span id="cb15-5"><a href="#cb15-5"></a><span class="co">#&gt; rowRanges(vcf):</span></span>
<span id="cb15-6"><a href="#cb15-6"></a><span class="co">#&gt;   GRanges with 5 metadata columns: paramRangeID, REF, ALT, QUAL, FILTER</span></span>
<span id="cb15-7"><a href="#cb15-7"></a><span class="co">#&gt; info(vcf):</span></span>
<span id="cb15-8"><a href="#cb15-8"></a><span class="co">#&gt;   DataFrame with 3 columns: AF, AC, AN</span></span>
<span id="cb15-9"><a href="#cb15-9"></a><span class="co">#&gt; info(header(vcf)):</span></span>
<span id="cb15-10"><a href="#cb15-10"></a><span class="co">#&gt;       Number Type    Description                                </span></span>
<span id="cb15-11"><a href="#cb15-11"></a><span class="co">#&gt;    AF A      Float   Allele Frequency                           </span></span>
<span id="cb15-12"><a href="#cb15-12"></a><span class="co">#&gt;    AC A      Integer Allele count in genotypes                  </span></span>
<span id="cb15-13"><a href="#cb15-13"></a><span class="co">#&gt;    AN 1      Integer Total number of alleles in called genotypes</span></span>
<span id="cb15-14"><a href="#cb15-14"></a><span class="co">#&gt; geno(vcf):</span></span>
<span id="cb15-15"><a href="#cb15-15"></a><span class="co">#&gt;   SimpleList of length 9: ES, SE, LP, AF, SS, EZ, SI, NC, ID</span></span>
<span id="cb15-16"><a href="#cb15-16"></a><span class="co">#&gt; geno(header(vcf)):</span></span>
<span id="cb15-17"><a href="#cb15-17"></a><span class="co">#&gt;       Number Type   Description                                           </span></span>
<span id="cb15-18"><a href="#cb15-18"></a><span class="co">#&gt;    ES A      Float  Effect size estimate relative to the alternative al...</span></span>
<span id="cb15-19"><a href="#cb15-19"></a><span class="co">#&gt;    SE A      Float  Standard error of effect size estimate                </span></span>
<span id="cb15-20"><a href="#cb15-20"></a><span class="co">#&gt;    LP A      Float  -log10 p-value for effect estimate                    </span></span>
<span id="cb15-21"><a href="#cb15-21"></a><span class="co">#&gt;    AF A      Float  Alternate allele frequency in the association study   </span></span>
<span id="cb15-22"><a href="#cb15-22"></a><span class="co">#&gt;    SS A      Float  Sample size used to estimate genetic effect           </span></span>
<span id="cb15-23"><a href="#cb15-23"></a><span class="co">#&gt;    EZ A      Float  Z-score provided if it was used to derive the EFFEC...</span></span>
<span id="cb15-24"><a href="#cb15-24"></a><span class="co">#&gt;    SI A      Float  Accuracy score of summary data imputation             </span></span>
<span id="cb15-25"><a href="#cb15-25"></a><span class="co">#&gt;    NC A      Float  Number of cases used to estimate genetic effect       </span></span>
<span id="cb15-26"><a href="#cb15-26"></a><span class="co">#&gt;    ID 1      String Study variant identifier</span></span></code></pre></div>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1"></a>vcf &lt;-<span class="st"> </span><span class="kw"><a href="../reference/query_gwas.html">query_gwas</a></span>(vcffile, <span class="dt">pval=</span><span class="fl">0.5</span>)</span>
<span id="cb16-2"><a href="#cb16-2"></a>vcf</span>
<span id="cb16-3"><a href="#cb16-3"></a><span class="co">#&gt; class: CollapsedVCF </span></span>
<span id="cb16-4"><a href="#cb16-4"></a><span class="co">#&gt; dim: 45 1 </span></span>
<span id="cb16-5"><a href="#cb16-5"></a><span class="co">#&gt; rowRanges(vcf):</span></span>
<span id="cb16-6"><a href="#cb16-6"></a><span class="co">#&gt;   GRanges with 5 metadata columns: paramRangeID, REF, ALT, QUAL, FILTER</span></span>
<span id="cb16-7"><a href="#cb16-7"></a><span class="co">#&gt; info(vcf):</span></span>
<span id="cb16-8"><a href="#cb16-8"></a><span class="co">#&gt;   DataFrame with 3 columns: AF, AC, AN</span></span>
<span id="cb16-9"><a href="#cb16-9"></a><span class="co">#&gt; info(header(vcf)):</span></span>
<span id="cb16-10"><a href="#cb16-10"></a><span class="co">#&gt;       Number Type    Description                                </span></span>
<span id="cb16-11"><a href="#cb16-11"></a><span class="co">#&gt;    AF A      Float   Allele Frequency                           </span></span>
<span id="cb16-12"><a href="#cb16-12"></a><span class="co">#&gt;    AC A      Integer Allele count in genotypes                  </span></span>
<span id="cb16-13"><a href="#cb16-13"></a><span class="co">#&gt;    AN 1      Integer Total number of alleles in called genotypes</span></span>
<span id="cb16-14"><a href="#cb16-14"></a><span class="co">#&gt; geno(vcf):</span></span>
<span id="cb16-15"><a href="#cb16-15"></a><span class="co">#&gt;   SimpleList of length 9: ES, SE, LP, AF, SS, EZ, SI, NC, ID</span></span>
<span id="cb16-16"><a href="#cb16-16"></a><span class="co">#&gt; geno(header(vcf)):</span></span>
<span id="cb16-17"><a href="#cb16-17"></a><span class="co">#&gt;       Number Type   Description                                           </span></span>
<span id="cb16-18"><a href="#cb16-18"></a><span class="co">#&gt;    ES A      Float  Effect size estimate relative to the alternative al...</span></span>
<span id="cb16-19"><a href="#cb16-19"></a><span class="co">#&gt;    SE A      Float  Standard error of effect size estimate                </span></span>
<span id="cb16-20"><a href="#cb16-20"></a><span class="co">#&gt;    LP A      Float  -log10 p-value for effect estimate                    </span></span>
<span id="cb16-21"><a href="#cb16-21"></a><span class="co">#&gt;    AF A      Float  Alternate allele frequency in the association study   </span></span>
<span id="cb16-22"><a href="#cb16-22"></a><span class="co">#&gt;    SS A      Float  Sample size used to estimate genetic effect           </span></span>
<span id="cb16-23"><a href="#cb16-23"></a><span class="co">#&gt;    EZ A      Float  Z-score provided if it was used to derive the EFFEC...</span></span>
<span id="cb16-24"><a href="#cb16-24"></a><span class="co">#&gt;    SI A      Float  Accuracy score of summary data imputation             </span></span>
<span id="cb16-25"><a href="#cb16-25"></a><span class="co">#&gt;    NC A      Float  Number of cases used to estimate genetic effect       </span></span>
<span id="cb16-26"><a href="#cb16-26"></a><span class="co">#&gt;    ID 1      String Study variant identifier</span></span></code></pre></div>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1"></a>vcf &lt;-<span class="st"> </span><span class="kw"><a href="../reference/query_gwas.html">query_gwas</a></span>(vcffile, <span class="dt">chrompos=</span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"1:1097291-1099437"</span>))</span>
<span id="cb17-2"><a href="#cb17-2"></a>vcf</span>
<span id="cb17-3"><a href="#cb17-3"></a><span class="co">#&gt; class: CollapsedVCF </span></span>
<span id="cb17-4"><a href="#cb17-4"></a><span class="co">#&gt; dim: 2 1 </span></span>
<span id="cb17-5"><a href="#cb17-5"></a><span class="co">#&gt; rowRanges(vcf):</span></span>
<span id="cb17-6"><a href="#cb17-6"></a><span class="co">#&gt;   GRanges with 5 metadata columns: paramRangeID, REF, ALT, QUAL, FILTER</span></span>
<span id="cb17-7"><a href="#cb17-7"></a><span class="co">#&gt; info(vcf):</span></span>
<span id="cb17-8"><a href="#cb17-8"></a><span class="co">#&gt;   DataFrame with 3 columns: AF, AC, AN</span></span>
<span id="cb17-9"><a href="#cb17-9"></a><span class="co">#&gt; info(header(vcf)):</span></span>
<span id="cb17-10"><a href="#cb17-10"></a><span class="co">#&gt;       Number Type    Description                                </span></span>
<span id="cb17-11"><a href="#cb17-11"></a><span class="co">#&gt;    AF A      Float   Allele Frequency                           </span></span>
<span id="cb17-12"><a href="#cb17-12"></a><span class="co">#&gt;    AC A      Integer Allele count in genotypes                  </span></span>
<span id="cb17-13"><a href="#cb17-13"></a><span class="co">#&gt;    AN 1      Integer Total number of alleles in called genotypes</span></span>
<span id="cb17-14"><a href="#cb17-14"></a><span class="co">#&gt; geno(vcf):</span></span>
<span id="cb17-15"><a href="#cb17-15"></a><span class="co">#&gt;   SimpleList of length 9: ES, SE, LP, AF, SS, EZ, SI, NC, ID</span></span>
<span id="cb17-16"><a href="#cb17-16"></a><span class="co">#&gt; geno(header(vcf)):</span></span>
<span id="cb17-17"><a href="#cb17-17"></a><span class="co">#&gt;       Number Type   Description                                           </span></span>
<span id="cb17-18"><a href="#cb17-18"></a><span class="co">#&gt;    ES A      Float  Effect size estimate relative to the alternative al...</span></span>
<span id="cb17-19"><a href="#cb17-19"></a><span class="co">#&gt;    SE A      Float  Standard error of effect size estimate                </span></span>
<span id="cb17-20"><a href="#cb17-20"></a><span class="co">#&gt;    LP A      Float  -log10 p-value for effect estimate                    </span></span>
<span id="cb17-21"><a href="#cb17-21"></a><span class="co">#&gt;    AF A      Float  Alternate allele frequency in the association study   </span></span>
<span id="cb17-22"><a href="#cb17-22"></a><span class="co">#&gt;    SS A      Float  Sample size used to estimate genetic effect           </span></span>
<span id="cb17-23"><a href="#cb17-23"></a><span class="co">#&gt;    EZ A      Float  Z-score provided if it was used to derive the EFFEC...</span></span>
<span id="cb17-24"><a href="#cb17-24"></a><span class="co">#&gt;    SI A      Float  Accuracy score of summary data imputation             </span></span>
<span id="cb17-25"><a href="#cb17-25"></a><span class="co">#&gt;    NC A      Float  Number of cases used to estimate genetic effect       </span></span>
<span id="cb17-26"><a href="#cb17-26"></a><span class="co">#&gt;    ID 1      String Study variant identifier</span></span></code></pre></div>
<p>It’s possible to chain filters together e.g.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1"></a>vcf &lt;-<span class="st"> </span><span class="kw"><a href="../reference/query_gwas.html">query_gwas</a></span>(vcffile, <span class="dt">rsid=</span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"rs3128126"</span>, <span class="st">"rs3121561"</span>, <span class="st">"rs3813193"</span>)) <span class="op">%&gt;%</span></span>
<span id="cb18-2"><a href="#cb18-2"></a><span class="st">    </span><span class="kw"><a href="../reference/query_gwas.html">query_gwas</a></span>(<span class="dt">pval=</span><span class="fl">0.5</span>)</span>
<span id="cb18-3"><a href="#cb18-3"></a>vcf</span>
<span id="cb18-4"><a href="#cb18-4"></a><span class="co">#&gt; class: CollapsedVCF </span></span>
<span id="cb18-5"><a href="#cb18-5"></a><span class="co">#&gt; dim: 1 1 </span></span>
<span id="cb18-6"><a href="#cb18-6"></a><span class="co">#&gt; rowRanges(vcf):</span></span>
<span id="cb18-7"><a href="#cb18-7"></a><span class="co">#&gt;   GRanges with 5 metadata columns: paramRangeID, REF, ALT, QUAL, FILTER</span></span>
<span id="cb18-8"><a href="#cb18-8"></a><span class="co">#&gt; info(vcf):</span></span>
<span id="cb18-9"><a href="#cb18-9"></a><span class="co">#&gt;   DataFrame with 3 columns: AF, AC, AN</span></span>
<span id="cb18-10"><a href="#cb18-10"></a><span class="co">#&gt; info(header(vcf)):</span></span>
<span id="cb18-11"><a href="#cb18-11"></a><span class="co">#&gt;       Number Type    Description                                </span></span>
<span id="cb18-12"><a href="#cb18-12"></a><span class="co">#&gt;    AF A      Float   Allele Frequency                           </span></span>
<span id="cb18-13"><a href="#cb18-13"></a><span class="co">#&gt;    AC A      Integer Allele count in genotypes                  </span></span>
<span id="cb18-14"><a href="#cb18-14"></a><span class="co">#&gt;    AN 1      Integer Total number of alleles in called genotypes</span></span>
<span id="cb18-15"><a href="#cb18-15"></a><span class="co">#&gt; geno(vcf):</span></span>
<span id="cb18-16"><a href="#cb18-16"></a><span class="co">#&gt;   SimpleList of length 9: ES, SE, LP, AF, SS, EZ, SI, NC, ID</span></span>
<span id="cb18-17"><a href="#cb18-17"></a><span class="co">#&gt; geno(header(vcf)):</span></span>
<span id="cb18-18"><a href="#cb18-18"></a><span class="co">#&gt;       Number Type   Description                                           </span></span>
<span id="cb18-19"><a href="#cb18-19"></a><span class="co">#&gt;    ES A      Float  Effect size estimate relative to the alternative al...</span></span>
<span id="cb18-20"><a href="#cb18-20"></a><span class="co">#&gt;    SE A      Float  Standard error of effect size estimate                </span></span>
<span id="cb18-21"><a href="#cb18-21"></a><span class="co">#&gt;    LP A      Float  -log10 p-value for effect estimate                    </span></span>
<span id="cb18-22"><a href="#cb18-22"></a><span class="co">#&gt;    AF A      Float  Alternate allele frequency in the association study   </span></span>
<span id="cb18-23"><a href="#cb18-23"></a><span class="co">#&gt;    SS A      Float  Sample size used to estimate genetic effect           </span></span>
<span id="cb18-24"><a href="#cb18-24"></a><span class="co">#&gt;    EZ A      Float  Z-score provided if it was used to derive the EFFEC...</span></span>
<span id="cb18-25"><a href="#cb18-25"></a><span class="co">#&gt;    SI A      Float  Accuracy score of summary data imputation             </span></span>
<span id="cb18-26"><a href="#cb18-26"></a><span class="co">#&gt;    NC A      Float  Number of cases used to estimate genetic effect       </span></span>
<span id="cb18-27"><a href="#cb18-27"></a><span class="co">#&gt;    ID 1      String Study variant identifier</span></span></code></pre></div>
<p>It’s possible to have multiple GWAS studies per vcf. You can specify specific GWAS studies to read in using e.g.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1"></a>vcf &lt;-<span class="st"> </span><span class="kw"><a href="../reference/query_gwas.html">query_gwas</a></span>(vcffile, <span class="dt">rsid=</span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"rs3128126"</span>, <span class="st">"rs3121561"</span>, <span class="st">"rs3813193"</span>), <span class="dt">id=</span><span class="st">"IEU-a-2"</span>)</span></code></pre></div>
<p>Note that querying by chrompos is the fastest way to deal with VCFs, use this over rsid where possible when speed is an issue.</p>
</div>
<div id="a-note-about-chrompos" class="section level2">
<h2 class="hasAnchor">
<a href="#a-note-about-chrompos" class="anchor"></a>A note about chrompos</h2>
<p>The fastest way to query VCFs is by specifying chromosome and position. Can specify specific positions, or ranges. e.g.</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1"></a>cp &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"1:10000"</span>, <span class="st">"2:10000-20000"</span>)</span></code></pre></div>
<p>or as a data frame</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1"></a>cp &lt;-<span class="st"> </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/reexports.html">tibble</a></span>(<span class="dt">chrom=</span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">1</span>,<span class="dv">2</span>), <span class="dt">start=</span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">10000</span>,<span class="dv">10000</span>), <span class="dt">end=</span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">10000</span>, <span class="dv">20000</span>))</span></code></pre></div>
<p>You can check what will be parsed out with:</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1"></a><span class="kw"><a href="../reference/parse_chrompos.html">parse_chrompos</a></span>(cp)</span>
<span id="cb22-2"><a href="#cb22-2"></a><span class="co">#&gt; GRanges object with 2 ranges and 0 metadata columns:</span></span>
<span id="cb22-3"><a href="#cb22-3"></a><span class="co">#&gt;       seqnames      ranges strand</span></span>
<span id="cb22-4"><a href="#cb22-4"></a><span class="co">#&gt;          &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt;</span></span>
<span id="cb22-5"><a href="#cb22-5"></a><span class="co">#&gt;   [1]        1       10000      *</span></span>
<span id="cb22-6"><a href="#cb22-6"></a><span class="co">#&gt;   [2]        2 10000-20000      *</span></span>
<span id="cb22-7"><a href="#cb22-7"></a><span class="co">#&gt;   -------</span></span>
<span id="cb22-8"><a href="#cb22-8"></a><span class="co">#&gt;   seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></span></code></pre></div>
<p>Querying by p-value or rsid is also possible but is slower as only chrompos is indexed. On Mac and Linux, rsid and p-value queries are performed by calls to bcftools. On Windows it uses VariantAnnotation directly, because bcftools binaries are not available. This is unfortunately somewhat slower. If many operations are being performed it might be faster to read in the whole dataset and perform queries that way.</p>
</div>
<div id="ld-proxies" class="section level2">
<h2 class="hasAnchor">
<a href="#ld-proxies" class="anchor"></a>LD proxies</h2>
<p>If a set of rsids are requested from a vcf but some are absent, a reference panel can be used to search for LD proxies, extract them, and align the effects and alleles against the original variants that were requested.</p>
<p>An LD reference panel can be obtained from here: <a href="http://fileserve.mrcieu.ac.uk/ld/data_maf0.01_rs_ref.tgz">http://fileserve.mrcieu.ac.uk/ld/data_maf0.01_rs_ref.tgz</a>. This dataset comprises Europeans from the 1000 genomes project, in plink format, and including only SNPs with MAF &gt; 0.01, and with the reference alleles aligned to the human genome reference sequence. For this vignette we can use a small subset of that dataset:</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1"></a>ldfile &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"eur.bed"</span>, <span class="dt">package=</span><span class="st">"gwasvcf"</span>) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb23-2"><a href="#cb23-2"></a><span class="st">  </span><span class="kw"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span>(<span class="st">".bed"</span>, <span class="st">""</span>, .)</span></code></pre></div>
<p>We also need to provide a path to the plink binary used to generate LD calculations. This can be done through the <code>genetics.binaRies</code> package as with bcftools</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1"></a><span class="kw"><a href="../reference/set_plink.html">set_plink</a></span>()</span>
<span id="cb24-2"><a href="#cb24-2"></a><span class="co">#&gt; Path not provided, using binaries in the explodecomputer/genetics.binaRies package</span></span></code></pre></div>
<p>The rs4442317 variant is not present in the vcf file, i.e. if we query that variant:</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1"></a><span class="kw"><a href="../reference/query_gwas.html">query_gwas</a></span>(vcffile, <span class="dt">rsid=</span><span class="st">"rs4442317"</span>) <span class="op">%&gt;%</span><span class="st"> </span>nrow</span>
<span id="cb25-2"><a href="#cb25-2"></a><span class="co">#&gt; [1] 0</span></span></code></pre></div>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1"></a>vcf &lt;-<span class="st"> </span><span class="kw"><a href="../reference/query_gwas.html">query_gwas</a></span>(vcffile, <span class="dt">rsid=</span><span class="st">"rs4442317"</span>, <span class="dt">proxies=</span><span class="st">"yes"</span>, <span class="dt">bfile=</span>ldfile, <span class="dt">tag_r2=</span><span class="fl">0.05</span>)</span>
<span id="cb26-2"><a href="#cb26-2"></a><span class="co">#&gt; Taking input= as a system command ('gunzip -c /var/folders/8t/4mzsfqzd5773rsc11w4q5fhw0000gp/T//RtmpuJJ49l/file16e7979fd5d16.targets.ld.gz') and a variable has been used in the expression passed to `input=`. Please use fread(cmd=...). There is a security concern if you are creating an app, and the app could have a malicious user, and the app is not running in a secure environment; e.g. the app is running as root. Please read item 5 in the NEWS file for v1.11.6 for more information and for the option to suppress this message.</span></span>
<span id="cb26-3"><a href="#cb26-3"></a>vcf <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/vcf_to_granges.html">vcf_to_granges</a></span>()</span>
<span id="cb26-4"><a href="#cb26-4"></a><span class="co">#&gt; GRanges object with 1 range and 15 metadata columns:</span></span>
<span id="cb26-5"><a href="#cb26-5"></a><span class="co">#&gt;             seqnames    ranges strand |            REF            ALT</span></span>
<span id="cb26-6"><a href="#cb26-6"></a><span class="co">#&gt;                &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;DNAStringSet&gt; &lt;DNAStringSet&gt;</span></span>
<span id="cb26-7"><a href="#cb26-7"></a><span class="co">#&gt;   rs4442317        1   1106784      * |              .              .</span></span>
<span id="cb26-8"><a href="#cb26-8"></a><span class="co">#&gt;                  QUAL      FILTER        ES        SE        LP        AF</span></span>
<span id="cb26-9"><a href="#cb26-9"></a><span class="co">#&gt;             &lt;numeric&gt; &lt;character&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span id="cb26-10"><a href="#cb26-10"></a><span class="co">#&gt;   rs4442317      &lt;NA&gt;        &lt;NA&gt;    0.0059    0.0071  0.391474    0.8559</span></span>
<span id="cb26-11"><a href="#cb26-11"></a><span class="co">#&gt;                    SS        EZ        SI        NC          ID</span></span>
<span id="cb26-12"><a href="#cb26-12"></a><span class="co">#&gt;             &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;character&gt;</span></span>
<span id="cb26-13"><a href="#cb26-13"></a><span class="co">#&gt;   rs4442317    138001      &lt;NA&gt;      &lt;NA&gt;      &lt;NA&gt;   rs4970420</span></span>
<span id="cb26-14"><a href="#cb26-14"></a><span class="co">#&gt;                      PR          id</span></span>
<span id="cb26-15"><a href="#cb26-15"></a><span class="co">#&gt;             &lt;character&gt; &lt;character&gt;</span></span>
<span id="cb26-16"><a href="#cb26-16"></a><span class="co">#&gt;   rs4442317   rs4970420     IEU-a-2</span></span>
<span id="cb26-17"><a href="#cb26-17"></a><span class="co">#&gt;   -------</span></span>
<span id="cb26-18"><a href="#cb26-18"></a><span class="co">#&gt;   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code></pre></div>
<p>Here we see that the proxy variant is rs4970420.</p>
<p>You may also extract only the best available proxies even if the requested rsids are present, by using <code>proxies="only"</code>. An example of this shows that the effect size estimates for the proxy variants are aligned to the effect alleles of the target variants:</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1"></a><span class="co"># Read vcf</span></span>
<span id="cb27-2"><a href="#cb27-2"></a>a &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/VariantAnnotation/man/readVcf-methods.html">readVcf</a></span>(vcffile)</span>
<span id="cb27-3"><a href="#cb27-3"></a></span>
<span id="cb27-4"><a href="#cb27-4"></a><span class="co"># Obtain the best LD proxy for each of the rsids</span></span>
<span id="cb27-5"><a href="#cb27-5"></a>b &lt;-<span class="st"> </span><span class="kw"><a href="../reference/query_gwas.html">query_gwas</a></span>(vcffile, <span class="dt">rsid=</span><span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(a), <span class="dt">proxies=</span><span class="st">"only"</span>, <span class="dt">bfile=</span>ldfile, <span class="dt">tag_r2=</span><span class="fl">0.6</span>)</span>
<span id="cb27-6"><a href="#cb27-6"></a><span class="co">#&gt; Taking input= as a system command ('gunzip -c /var/folders/8t/4mzsfqzd5773rsc11w4q5fhw0000gp/T//RtmpuJJ49l/file16e797a67fa81.targets.ld.gz') and a variable has been used in the expression passed to `input=`. Please use fread(cmd=...). There is a security concern if you are creating an app, and the app could have a malicious user, and the app is not running in a secure environment; e.g. the app is running as root. Please read item 5 in the NEWS file for v1.11.6 for more information and for the option to suppress this message.</span></span>
<span id="cb27-7"><a href="#cb27-7"></a></span>
<span id="cb27-8"><a href="#cb27-8"></a><span class="co"># Match the target data to the proxy data</span></span>
<span id="cb27-9"><a href="#cb27-9"></a>index &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/match.html">match</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(b), <span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(a))</span>
<span id="cb27-10"><a href="#cb27-10"></a></span>
<span id="cb27-11"><a href="#cb27-11"></a><span class="co"># Plot the target data effects against the proxy data effects</span></span>
<span id="cb27-12"><a href="#cb27-12"></a><span class="kw"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span>(<span class="kw"><a href="../reference/vcf_to_granges.html">vcf_to_granges</a></span>(b)<span class="op">$</span>ES, <span class="kw"><a href="../reference/vcf_to_granges.html">vcf_to_granges</a></span>(a)<span class="op">$</span>ES[index])</span></code></pre></div>
<p><img src="guide_files/figure-html/unnamed-chunk-23-1.png" width="700"></p>
</div>
<div id="creating-the-vcf-object-from-a-data-frame" class="section level2">
<h2 class="hasAnchor">
<a href="#creating-the-vcf-object-from-a-data-frame" class="anchor"></a>Creating the VCF object from a data frame</h2>
<p>If you have GWAS summary data in a text file or data frame, this can be converted to a VCF object.</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1"></a>vcf &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/VariantAnnotation/man/readVcf-methods.html">readVcf</a></span>(vcffile)</span>
<span id="cb28-2"><a href="#cb28-2"></a>vv &lt;-<span class="st"> </span><span class="kw"><a href="../reference/vcf_to_granges.html">vcf_to_granges</a></span>(vcf) <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/reexports.html">as_tibble</a></span>()</span>
<span id="cb28-3"><a href="#cb28-3"></a>out &lt;-<span class="st"> </span>vv <span class="op">%$%</span><span class="st"> </span><span class="kw"><a href="../reference/create_vcf.html">create_vcf</a></span>(<span class="dt">chrom=</span>seqnames, <span class="dt">pos=</span>start, <span class="dt">nea=</span>REF, <span class="dt">ea=</span>ALT, <span class="dt">snp=</span>ID, <span class="dt">ea_af=</span>AF, <span class="dt">effect=</span>ES, <span class="dt">se=</span>SE, <span class="dt">pval=</span><span class="dv">10</span><span class="op">^-</span>LP, <span class="dt">n=</span>SS, <span class="dt">name=</span><span class="st">"a"</span>)</span>
<span id="cb28-4"><a href="#cb28-4"></a>out</span>
<span id="cb28-5"><a href="#cb28-5"></a><span class="co">#&gt; class: CollapsedVCF </span></span>
<span id="cb28-6"><a href="#cb28-6"></a><span class="co">#&gt; dim: 92 1 </span></span>
<span id="cb28-7"><a href="#cb28-7"></a><span class="co">#&gt; rowRanges(vcf):</span></span>
<span id="cb28-8"><a href="#cb28-8"></a><span class="co">#&gt;   GRanges with 4 metadata columns: REF, ALT, QUAL, FILTER</span></span>
<span id="cb28-9"><a href="#cb28-9"></a><span class="co">#&gt; info(vcf):</span></span>
<span id="cb28-10"><a href="#cb28-10"></a><span class="co">#&gt;   DataFrame with 0 columns: </span></span>
<span id="cb28-11"><a href="#cb28-11"></a><span class="co">#&gt; geno(vcf):</span></span>
<span id="cb28-12"><a href="#cb28-12"></a><span class="co">#&gt;   SimpleList of length 5: AF, ES, SE, LP, SS</span></span>
<span id="cb28-13"><a href="#cb28-13"></a><span class="co">#&gt; geno(header(vcf)):</span></span>
<span id="cb28-14"><a href="#cb28-14"></a><span class="co">#&gt;       Number Type  Description                                            </span></span>
<span id="cb28-15"><a href="#cb28-15"></a><span class="co">#&gt;    AF A      Float Alternate allele frequency in the association study    </span></span>
<span id="cb28-16"><a href="#cb28-16"></a><span class="co">#&gt;    ES A      Float Effect size estimate relative to the alternative allele</span></span>
<span id="cb28-17"><a href="#cb28-17"></a><span class="co">#&gt;    SE A      Float Standard error of effect size estimate                 </span></span>
<span id="cb28-18"><a href="#cb28-18"></a><span class="co">#&gt;    LP A      Float -log10 p-value for effect estimate                     </span></span>
<span id="cb28-19"><a href="#cb28-19"></a><span class="co">#&gt;    SS A      Float Sample size used to estimate genetic effect</span></span></code></pre></div>
<p>It’s possible to write the vcf file:</p>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1"></a><span class="kw"><a href="https://rdrr.io/pkg/VariantAnnotation/man/writeVcf-methods.html">writeVcf</a></span>(out, <span class="dt">file=</span><span class="st">"temp.vcf"</span>)</span></code></pre></div>
<p>You may want to first harmonise the data so that all the non-effect alleles are aligned to the human genome reference. See the <a href="https://github.com/MRCIEU/gwasglue">gwasglue</a> package on some functions to do this.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#external-tools">External tools</a></li>
      <li><a href="#reading-in-everything">Reading in everything</a></li>
      <li><a href="#converting-to-simple-dataframes">Converting to simple dataframes</a></li>
      <li><a href="#reading-in-with-filters">Reading in with filters</a></li>
      <li><a href="#a-note-about-chrompos">A note about chrompos</a></li>
      <li><a href="#ld-proxies">LD proxies</a></li>
      <li><a href="#creating-the-vcf-object-from-a-data-frame">Creating the VCF object from a data frame</a></li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Gibran Hemani.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
